### PR TITLE
libflux: clean up bulk message handler registration functions

### PR DESCRIPTION
**Pull request by garlick**
This is following up on an idea @morrone had in #1135 to clean up the bulk message handler reg/unreg functions.  The `flux_msg_handler_t` member is dropped from `struct flux_msg_handler_spec`, and instead a new array of message handlers is returned from `flux_msg_handler_addvec()`.

`flux_msg_handler_delvec()` then accepts the array of message handlers, destroying them and freeing the array.

In addition, a small mention about rolemask security is added to the message handler man pages.

Updated users, tests, and man pages.

This will require a PR to flux-sched.  Please don't merge until that's queued up.

### CODE REVIEW

**Comment by garlick** [comment]
Yes, I'll post a sched PR once @morrone chimes in here.

**Comment by grondo** [comment]
Looks good to me as well.  I would have leaned toward renaming the functions to something more like create/destroy, but after going through the code I think addvec/delvec works just as well so probably a better choice to keep the existing naming.

**Comment by garlick** [comment]
Thanks disembodied voice of @grondo from vacation :-)

File: src/common/libflux/msg_handler.c [hunk 1]
```diff
@@ -556,51 +556,64 @@ flux_msg_handler_t *flux_msg_handler_create (flux_t *h,
...
-        i--;
-    }
+    saved_errno = errno;
+    flux_msg_handler_delvec (handlers);
+    errno = saved_errno;
     return -1;
 }
 
-void flux_msg_handler_delvec (struct flux_msg_handler_spec tab[])
+void flux_msg_handler_delvec (flux_msg_handler_t **handlers)
```
Comment (morrone) [inline src/common/libflux/msg_handler.c]: *handlers[]?

File: src/broker/shutdown.c [hunk 2]
```diff
@@ -63,6 +63,7 @@ void shutdown_destroy (shutdown_t *s)
     if (s) {
         if (s->shutdown)
             flux_msg_handler_destroy (s->shutdown);
+        shutdown_disarm (s);
```
Comment (morrone) [inline src/broker/shutdown.c]: Perhaps a good change, but unrelated to the logical topic of this commit.

File: src/common/libflux/msg_handler.c [hunk 3]
```diff
@@ -556,51 +556,64 @@ flux_msg_handler_t *flux_msg_handler_create (flux_t *h,
...
+    return (spec.typemask == end.typemask
+            && spec.topic_glob == end.topic_glob
+            && spec.cb == end.cb
+            && spec.rolemask == end.rolemask);
+}
+
+int flux_msg_handler_addvec (flux_t *h,
+                             const struct flux_msg_handler_spec tab[],
+                             void *arg,
+                             flux_msg_handler_t ***hp)
```
Comment (morrone) [inline src/common/libflux/msg_handler.c]: Perhaps "flux_msg_handler_t **hp[]" would make reader comprehension a bit easier?

File: src/common/libflux/msg_handler.c [hunk 4]
```diff
@@ -556,51 +556,64 @@ flux_msg_handler_t *flux_msg_handler_create (flux_t *h,
...
+    flux_msg_handler_t **handlers = NULL;
+    int count = 0;
+    int saved_errno;
 
-    for (i = 0; ; i++) {
-        if (!tab[i].typemask && !tab[i].topic_glob && !tab[i].cb)
-            break; /* FLUX_MSGHANDLER_TABLE_END */
+    if (!h || !tab || !hp) {
+        errno = EINVAL;
+        goto error;
```
Comment (morrone) [inline src/common/libflux/msg_handler.c]: There would appear to be no reason to call flux_msg_handler_delvec() if this memory allocation fails.  This should probably jump to another label just before "return -1".

File: src/common/libflux/msg_handler.c [hunk 5]
```diff
@@ -556,51 +556,64 @@ flux_msg_handler_t *flux_msg_handler_create (flux_t *h,
...
+    }
+    while (!at_end (tab[count]))
+        count++;
+    if (!(handlers = calloc (count + 1, sizeof (flux_msg_handler_t *))))
+        goto error;
```
Comment (morrone) [inline src/common/libflux/msg_handler.c]: This shouldn't call flux_msg_handler_delvec() on error either.

File: src/common/libflux/msg_handler.c [hunk 6]
```diff
@@ -556,51 +556,64 @@ flux_msg_handler_t *flux_msg_handler_create (flux_t *h,
...
-    for (i = 0; ; i++) {
-        if (!tab[i].typemask && !tab[i].topic_glob && !tab[i].cb)
-            break; /* FLUX_MSGHANDLER_TABLE_END */
-        if (tab[i].w) {
-            flux_msg_handler_stop (tab[i].w);
-            flux_msg_handler_destroy (tab[i].w);
-            tab[i].w = NULL;
+    if (handlers) {
+        int i;
+        for (i = 0; handlers[i] != NULL; i++) {
```
Comment (morrone) [inline src/common/libflux/msg_handler.c]: Since we are using C99, I would prefer to see this be "for (int i =0;" rather than declare i on its own line.

File: src/common/libjsc/jstatctl.c [hunk 7]
```diff
@@ -1144,19 +1147,19 @@ static int notify_status_obj (flux_t *h, jsc_handler_obj_f func, void *d)
         rc = -1;
         goto done;
     }
-    if (flux_msg_handler_addvec (h, htab, (void *)ctx) < 0) {
+    if (flux_msg_handler_addvec (h, htab, NULL, &handlers) < 0) {
```
Comment (morrone) [inline src/common/libjsc/jstatctl.c]: The elimination of (void *)ctx doesn't appear to be related to this logical commit.

File: src/modules/wreck/job.c [hunk 8]
```diff
@@ -608,20 +608,23 @@ static void runevent_cb (flux_t *h, flux_msg_handler_t *w,
...
 int mod_main (flux_t *h, int argc, char **argv)
 {
-    if (flux_msg_handler_addvec (h, mtab, NULL) < 0) {
+    flux_msg_handler_t **handlers = NULL;
+    int rc = -1;
+
+    if (flux_msg_handler_addvec (h, mtab, NULL, &handlers) < 0) {
         flux_log_error (h, "flux_msg_handler_addvec");
-        return (-1);
+        goto done;
```
Comment (morrone) [inline src/modules/wreck/job.c]: This jumps to the wrong place.  If flux_msg_handler_addvec() failed, we shouldn't be calling flux_msg_handler_delvec().

File: src/broker/attr.c [hunk 9]
```diff
@@ -415,27 +416,24 @@ void lsattr_request_cb (flux_t *h, flux_msg_handler_t *mh,
  ** Initialization
  **/
 
-static struct flux_msg_handler_spec handlers[] = {
-    { FLUX_MSGTYPE_REQUEST, "attr.get",    getattr_request_cb, 0, NULL },
-    { FLUX_MSGTYPE_REQUEST, "attr.list",   lsattr_request_cb, 0, NULL },
-    { FLUX_MSGTYPE_REQUEST, "attr.set",    setattr_request_cb, 0, NULL },
+static const struct flux_msg_handler_spec handlers[] = {
+    { FLUX_MSGTYPE_REQUEST, "attr.get",    getattr_request_cb, FLUX_ROLE_ALL },
+    { FLUX_MSGTYPE_REQUEST, "attr.list",   lsattr_request_cb, FLUX_ROLE_ALL },
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/attr.c]: That's a hard one to split out.  The code was setting the rolemask to 0, then changing it with code like
```c
flux_msg_handler_allow_rolemask (handlers[0].w, FLUX_ROLE_ALL);
```
Since `w` is gone, I thought it best to change the initial setting rather than walk the `handlers` list or whatever.

File: src/broker/broker.c [hunk 10]
```diff
@@ -668,6 +669,8 @@ int main (int argc, char *argv[])
     service_switch_destroy (ctx.services);
     hello_destroy (ctx.hello);
     attr_destroy (ctx.attrs);
+    shutdown_destroy (ctx.shutdown);
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/broker.c]: Agreed, will split out.

File: src/broker/broker.c [hunk 11]
```diff
@@ -668,6 +669,8 @@ int main (int argc, char *argv[])
...
+    flux_msg_handler_delvec (ctx.handlers);
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/broker.c]: Good suggestion.

File: src/broker/shutdown.c (hunk 2 above)
Comment (garlick) [reply src/broker/shutdown.c]: Yep.

File: src/common/libflux/msg_handler.c (hunk 3 above)
Comment (garlick) [reply src/common/libflux/msg_handler.c]: Sure.

File: src/common/libflux/msg_handler.c (hunk 4 above)
Comment (garlick) [reply src/common/libflux/msg_handler.c]: OK.

File: src/common/libflux/msg_handler.c (hunk 5 above)
Comment (garlick) [reply src/common/libflux/msg_handler.c]: Sure

File: src/common/libflux/msg_handler.c (hunk 1 above)
Comment (garlick) [reply src/common/libflux/msg_handler.c]: OK

File: src/common/libflux/msg_handler.c (hunk 6 above)
Comment (garlick) [reply src/common/libflux/msg_handler.c]: OK...

File: src/common/libjsc/jstatctl.c (hunk 7 above)
Comment (garlick) [reply src/common/libjsc/jstatctl.c]: Yes, will split out.

File: src/modules/wreck/job.c (hunk 8 above)
Comment (garlick) [reply src/modules/wreck/job.c]: delvec is a no-op so it's not wrong per se.  Sometimes it's nice to have a single exit path rather than returns sprinkled throughout a function.    I could go either way here but I'll do as you suggest.

**Comment by garlick** [comment]
> Mainly, I would like to see the "cleanup addvec/delvec" commit reworked to only deal with the logical changes to separate the flux_msg_handler_t * member from flux_msg_handler_spec. Other logical changes should be in other commits.

I can't separate the libflux/msg_handler changes from the rest without breaking git bisect, so unfortunately it's a long commit, but it'll be a bit more focused after I split out the other bits you've suggested.

Thanks for the detailed feedback!

**Comment by morrone** [comment]
> I can't separate the libflux/msg_handler changes from the rest without breaking git bisect

I'm not entirely sure what you think I was asking for, but I don't think it was that. :)  I was viewing "separate the flux_msg_handler_t * member from flux_msg_handler_spec" as the one big change in that commit.  There was no additional implication meant to "separate the libflux/msg_handler changes from the rest".

**Comment by garlick** [comment]
Just pushed an update with the suggested changes.

File: doc/man3/flux_msg_handler_addvec.adoc [hunk 12]
```diff
@@ -49,6 +49,17 @@ handler, all previously created message handlers in the array are destroyed
 before an error is returned.
 
 
+SECURITY
+--------
+
+The _rolemask_ member of _struct flux_msg_handler_spec_ is passed
+to `flux_msg_handler_allow_rolemask()`.  If only FLUX_ROLE_OWNER should
+be allowed to send messages to the handler, set _rolemask_ to zero.
```
Comment (morrone) [inline doc/man3/flux_msg_handler_addvec.adoc]: I find this very confusing.  FLUX_ROLE_NONE  is 0, and FLUX_ROLE_OWNER is 1.  So essentially you are saying that if we want to allow FLUX_ROLE_OWNER, we should set FLUX_ROLE_NONE.

I'm not a fan of using zero in the code when we have explicitly set a name for that, FLUX_ROLE_NONE.  So at the least, I would prefer to see "zero" changed to "FLUX_ROLE_NONE".

I read the flux_msg_handler_create(3) man page as instructed below, and it says:

"A message handler only receives a message if a bit-wise AND of the message
rolemask and the message handler rolemask evaluates to a non-zero value.
By default, the message handler rolemask contains only FLUX_ROLE_OWNER
(the instance owner) but may be changed with
`flux_msg_handler_allow_rolemask()` and `flux_msg_handler_deny_rolemask()`,
which add or remove roles from the message handler rolemask.  FLUX_ROLE_OWNER
cannot be dropped from the message handler rolemask."

It is pretty unclear about what happens when we try to set the role to FLUX_ROLE_NONE, despite that being exactly what is instructed in this man page.  I'm assuming that it means that FLUX_ROLE_NONE is ignored and the default value of FLUX_ROLE_OWNER remains.

As I can tell from the documentation, at current time there is no practical difference between FLUX_ROLE_NONE and FLUX_ROLE_OWNER.  But perhaps in the future there will be? (Otherwise, why have them both?)  So perhaps better than setting it to 0/FLUX_ROLE_NONE, it would be better to instruct people to set FLUX_ROLE_OWNER if they want FLUX_ROLE_OWNER.

File: src/broker/broker.c [hunk 13]
```diff
@@ -158,7 +159,7 @@ static void broker_handle_signals (broker_ctx_t *ctx, zlist_t *sigwatchers);
 static void broker_unhandle_signals (zlist_t *sigwatchers);
 
 static void broker_add_services (broker_ctx_t *ctx);
-static void broker_remove_services (void);
+static void broker_remove_services (flux_msg_handler_t *handlers[]);
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/broker.c]: That's reasonable.

[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/broker.c]: > I'm confused about why that is hard

Oh duh, not sure what I was thinking there.  You are right.

File: doc/man3/flux_msg_handler_addvec.adoc (hunk 12 above)
Comment (garlick) [reply doc/man3/flux_msg_handler_addvec.adoc]: I'm pondering this one.

In general, I don't see a problem with 0 as a value for a bitmask as opposed to a symbolic value.  FLUX_ROLE_NONE was not really intended to be used here; it was defined to be used as the initial value for the rolemask field in a message, e.g. from [RFC 11](https://github.com/flux-framework/rfc/blob/master/spec_12.adoc)

> FLUX_ROLE_NONE (0) SHALL indicate "invalid rolemask".

Possibly that was a poor naming choice.

I think you are right that the way FLUX_ROLE_OWNER as described is confusing.  It cannot be denied and is meaningless to allow, so perhaps instead of being described as implicitly set in the "allow mask" of a handler, it should be an error to include it in the mask or to attempt to allow/deny it.  It would be unfortunate to give the impression that the owner can be denied, since the fact that the owner is all powerful is a fundamental part of the security design.

In any case this pr is about something else so maybe the best thing is to drop the man page update from this PR and do another.

**Comment by garlick** [comment]
security documentation dropped for another PR, and requested change to broker_add_services() added.

**Comment by morrone** [comment]
Looks pretty good.  Just missing the part in broker.c about keeping the "handlers" variable local to main() rather than more globally in ctx where it is not needed.

**Comment by garlick** [comment]
I went ahead and squashed that last change (where handlers is declared).

**Comment by garlick** [comment]
Got a buildbot failure after that rebase:  a recurrence of #1103 proxy event distribution (probably a racy test) seems to be the cause.  There are some false positives reported by cppcheck there too, but they were there in the last build that succeeded so I guess they do not affect overall pass/fail of the build?

I would rerun the build but not sure how to do that yet...
